### PR TITLE
[WIP RLlib] Enable A3C torch async and fix race-condition related to async `model.value_function()` calls.

### PR DESCRIPTION
**Pull request by sven1977**
Re-enable A3C torch async.
Thereby: Fix race-condition related to async `model.value_function()` calls.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by sven1977** [comment]
Waiting for all tests to pass.

File: rllib/models/modelv2.py [hunk 1]
```diff
@@ -189,12 +192,16 @@ def __call__(
             state (list): list of state tensors with sizes matching those
                 returned by get_initial_state + the batch dimension
             seq_lens (Tensor): 1d tensor holding input sequence lengths
+            return_values (bool): Whether to also return the values tensor.
+                If True, will return 3 items: [output, states, values], instead
+                of 2.
 
         Returns:
-            (outputs, state): The model output tensor of size
+            (outputs, state, values?): The model output tensor of size
```
Comment (roireshef) [inline rllib/models/modelv2.py:200]: I'd suggest leaving the shape of the tuple fixed, and either populate values or not based on that flag (maybe require the user to return values=None if flag is set to false?)

File: rllib/models/modelv2.py [hunk 2]
```diff
@@ -175,7 +176,9 @@ def __call__(
             self,
             input_dict: Dict[str, TensorType],
             state: List[Any] = None,
-            seq_lens: TensorType = None) -> (TensorType, List[TensorType]):
+            seq_lens: TensorType = None,
+            return_values: bool = False,
+    ) -> (TensorType, List[TensorType]):
```
Comment (roireshef) [inline rllib/models/modelv2.py:181]: the return part of the signature should probably change here

File: rllib/models/modelv2.py [hunk 3]
```diff
@@ -222,15 +230,20 @@ def __call__(
...
+        if return_values:
+            return outputs, state_outs, res[2]
+        return outputs, state_outs
 
     @PublicAPI
     def from_batch(self, train_batch: SampleBatch,
-                   is_training: bool = True) -> (TensorType, List[TensorType]):
+                   is_training: bool = True,
+                   return_values: bool = False) -> \
+            (TensorType, List[TensorType]):
```
Comment (roireshef) [inline rllib/models/modelv2.py:246]: signature (as above)

**Review by roireshef — COMMENTED** [review]
Hi @sven1977 - that was quick! I've added some comments on your PR

**Comment by ericl** [comment]
@sven1977 I'm paging this conversation back in; wasn't the conclusion of the previous conversation that there wasn't any real race condition?

The sequence of events hypothesized to lead to a race condition is this:
1. model.forward() is called. This stores model._value
2. model.set_weights() is called. This updates the model weights.
3. model.value_function() is called to retrieve model._value.

I don't think there is any issue here since model._value is never overwritten.

**Comment by ericl** [comment]
I see the current fcnet is not quite doing that, perhaps you can try this (which doesn't break any APIs):

```
diff --git a/rllib/models/torch/fcnet.py b/rllib/models/torch/fcnet.py
index 4080c9318..81a8c08ac 100644
--- a/rllib/models/torch/fcnet.py
+++ b/rllib/models/torch/fcnet.py
@@ -109,6 +109,8 @@ class FullyConnectedNetwork(TorchModelV2, nn.Module):
         self._features = None
         # Holds the last input, in case value branch is separate.
         self._last_flat_in = None
+        # Holds the current "value" branch output.
+        self._value = None
 
     @override(TorchModelV2)
     def forward(self, input_dict, state, seq_lens):
@@ -119,13 +121,14 @@ class FullyConnectedNetwork(TorchModelV2, nn.Module):
             self._features
         if self.free_log_std:
             logits = self._append_free_log_std(logits)
+        if self._value_branch_separate:
+            self._value = self._value_branch(
+                self._value_branch_separate(self._last_flat_in)).squeeze(1)
+        else:
+            self._value = self._value_branch(self._features).squeeze(1)
         return logits, state
 
     @override(TorchModelV2)
     def value_function(self):
-        assert self._features is not None, "must call forward() first"
-        if self._value_branch_separate:
-            return self._value_branch(
-                self._value_branch_separate(self._last_flat_in)).squeeze(1)
-        else:
-            return self._value_branch(self._features).squeeze(1)
+        assert self._value is not None, "must call forward() first"
+        return self._value
```

**Comment by roireshef** [comment]
hi @ericl - did you my last few comments on: https://github.com/ray-project/ray/issues/10628 ? 
I understand you say that theoretically it shouldn't happen, but it looks like it does. It keeps reproducing on my machine and I debugged it down to what seems to look like a race condition. I'd appreciate it if you could try reproducing that on your machine or contradict my conclusions with some other explanation to why do I see the "._value" being overridden

**Comment by sven1977** [comment]
Hm, In case the value branch is separate (and the values are not needed), wouldn't the above solution cause a lot of overhead in each forward pass?

**Comment by sven1977** [comment]
I agree changing the API is ugly.
Well, let's first figure out whether any of these solutions actually fixes Roi's problem with A3C torch.

**Comment by roireshef** [comment]
> Hm, In case the value branch is separate (and the values are not needed), wouldn't the above solution cause a lot of overhead in each forward pass?

I thought about an idea:
1. forward() can return everything but the value, but instead some intermediate result that can be used to compute the value (last tensor before the value head)
2. forward_with_value() can then wrap value and call the value head using intermediate results from forward()

Any implementation can then choose to call either one of the functions at any given time. 

WDYT?

**Comment by roireshef** [comment]
> I agree changing the API is ugly.
> Well, let's first figure out whether any of these solutions actually fixes Roi's problem with A3C torch.

Hi @sven1977 I understand the motivation behind minimizing API changes, but from my experience if one is needed, it's better to apply it sooner than later, because bad interfaces can make issues accumulate. 

In this case, if you still aiming to support PyTorch as an underlying infrastructure, the current API just doesn't make a lot of sense because it creates execution issues. The issues I'm talking about should be easy to reproduce...

What alternative do you find better? To the best of my understanding, leaving the API as is means PyTorch is not well supported anymore, which is a regression from the 0.7.x versions... Is it preferable to API changes?

**Comment by roireshef** [comment]
@ericl @sven1977 I have few insights. So far it looks like running the proposed PR code (I had to made some minor changes for it to work) is able to train with AsyncSampler and PyTorch well. Speedups are very noticeable compared to using SyncSampler, as expected.

I went ahead and modified @sven1977's PR to get the following PR: https://github.com/ray-project/ray/pull/12030
It basically has:
* ModelV2.forward() always returns 3 values: outputs, new_(rnn)_state, extra_ouputs. The last one being a dictionary that maps output name (str) to its corresponding tensor (same type as the "batch" objects). This gives the developer the flexibility to return any intermediate tensor (or additional head's outputs, such as value estimate, but not limited to only it) so it can be queried and used later on within the Agent's implementation, for example when computing its loss, such as in A3C. BTW a model that has no extra outputs can return a None value, similar to the case of returning an RNN state when not using RNN at all.

* TorchPolicy._compute_action_helper() peeks at the new extra_outputs returned value and if it is not empty, it merges it to its returned values (specifically "extra_fetches"). This method already computes few other additional values that are stored as SampleBatch fields, so this change doesn't change this method's scope or signature.

From that moment on, value function estimates are included as part of the "post-processed" batch values. 

This ultimately achieves few two things:
1. The same caching is achieved just like the current implementation in master
2. All the values returned by the ModelV2 are atomic, and are robust to the race conditions they are prone to now when using PyTorch. Specifically, AsyncSampler can be used by PyTorch users now.
3. Personally I think this is also a good opportunity to replace the ModelV2.value_function() with a more general solution, as now the user is able to return not just the value function estimate, but any arbitrary tensor of choice.
4. From a user's perspective, I believe this is as easy and concise to implement as implementing ModelV2.value_function method, if not better...

WDYT?

**Comment by ericl** [comment]
A couple comments.

1. I agree that it may make sense to evolve the ModelV2 API to return outputs immediately. However, this is a rather big change, we need to do it holistically and plan for backwards compatibility issues. Just patching it for torch/value_function makes it inconsistent with how other model output heads work.

2. We still don't know why there is a race condition, a tensor returned immediately vs a tensor saved and then returned a bit later in a method should have identical thread-safety behavior from pytorch's perspective. Have you tried out the patch from https://github.com/ray-project/ray/pull/11935#issuecomment-725615302 ?

**Comment by sven1977** [comment]
Yes, we do understand the race condition: It happens only in A3C, because of the "race" there between action computations (async samplers within RolloutWorker using the Policy object) and gradient calculations happening in the same RolloutWorker (and using the same Policy/Model!). This explains the "batch-size mismatch" errors (batches have different sizes for action calculations vs gradient calculation passes)

Ok, I think the following is a viable solution for right now, not breaking the ModelV2 API.
- Fix the problem only for A3C torch for now.
- Allow passing in a "return_value_estimate" flag into a Model's constructor, which is respected only by Torch FC/CNN/auto-LSTM and causes the model to return the value estimates as a third item in the returned tuple.

**Comment by roireshef** [comment]
> A couple comments.
> 
>     1. I agree that it may make sense to evolve the ModelV2 API to return outputs immediately. However, this is a rather big change, we need to do it holistically and plan for backwards compatibility issues. Just patching it for torch/value_function makes it inconsistent with how other model output heads work.
> 
>     2. We still don't know why there is a race condition, a tensor returned immediately vs a tensor saved and then returned a bit later in a method should have identical thread-safety behavior from pytorch's perspective. Have you tried out the patch from [#11935 (comment)](https://github.com/ray-project/ray/pull/11935#issuecomment-725615302) ?

Hi @ericl - please see https://github.com/ray-project/ray/pull/11935#issuecomment-727614473 and later comments for reference. Short answer is **yes**, and it solves the issue.

**Comment by roireshef** [comment]
@sven1977 - last commits have "WIP" in title. Is this PR ready? Should I try running it?

**Comment by ericl** [comment]
Discussed this offline with @sven1977 . We are going to do the following:
1. To solve the race condition, add lock annotations around class methods in TorchPolicy. This is a simpler fix that doesn't require any tricky changes for now.
2. Separately, start a design to allow direct branch output returns from ModelV2. This will be designed to be backwards compatible with existing ModelV2 models. We can start porting Torch/TF ModelV2 implementations to this new method if they don't need the performance benefits of separate fetch calls.

**Comment by roireshef** [comment]
> Discussed this offline with @sven1977 . We are going to do the following:
> 
>     1. To solve the race condition, add lock annotations around class methods in TorchPolicy. This is a simpler fix that doesn't require any tricky changes for now.
> 
>     2. Separately, start a design to allow direct branch output returns from ModelV2. This will be designed to be backwards compatible with existing ModelV2 models. We can start porting Torch/TF ModelV2 implementations to this new method if they don't need the performance benefits of separate fetch calls.

@ericl + @sven1977 - Thanks! Could you please update here (or LMK) once the change in (1) is merged to master? It is also worth removing the warning and flag update in A3C's **validate_config** function.

**Comment by roireshef** [comment]
> Discussed this offline with @sven1977 . We are going to do the following:
> 
>     1. To solve the race condition, add lock annotations around class methods in TorchPolicy. This is a simpler fix that doesn't require any tricky changes for now.
> 
>     2. Separately, start a design to allow direct branch output returns from ModelV2. This will be designed to be backwards compatible with existing ModelV2 models. We can start porting Torch/TF ModelV2 implementations to this new method if they don't need the performance benefits of separate fetch calls.

Hi @ericl / @sven1977 - I didn't understand if this is merged to master already or WIP...? Can you provide with the PR id/link so I can track it, since this PR is no longer open?

**Comment by roireshef** [comment]
@sven1977 @ericl is the fix already merged to master?

**Comment by roireshef** [comment]
>     1. To solve the race condition, add lock annotations around class methods in TorchPolicy. This is a simpler fix that doesn't require any tricky changes for now.

@sven1977 @ericl is this fix already merged to master?
